### ISSUE TITLE
nginx-proxy on ECS: default.conf missing upstream configuration; used to work

### ISSUE THREAD

**Issue by landon9720** [nginx-proxy/nginx-proxy#1004]
My organization uses nginx-proxy with nodejs running in AWS ECS. This configuration has worked reliably for many months. However, last week, some (but not all) of the nodes stopped handling health check requests, and the load balancer removed them from service. On investigation, I found that nginx-proxy was failing to start. The nginx-proxy log output is similar to the following:

```
$ docker logs 0742fa255f00
dockergen.1 | 2017/12/18 19:50:23 Generated '/etc/nginx/conf.d/default.conf' from 3 containers
dockergen.1 | 2017/12/18 19:50:23 Error running notify command: nginx -s reload, exit status 1
dockergen.1 | 2017/12/18 19:50:23 Watching docker events
dockergen.1 | 2017/12/18 19:50:23 Contents of /etc/nginx/conf.d/default.conf did not change. Skipping notification 'nginx -s reload'
2017/12/18 19:55:32 [emerg] 48#48: no servers are inside upstream in /etc/nginx/conf.d/default.conf:55
nginx: [emerg] no servers are inside upstream in /etc/nginx/conf.d/default.conf:55\
```

Here is the docker.conf generated by nginx-proxy:

```
# If we receive X-Forwarded-Proto, pass it through; otherwise, pass along the
# scheme used to connect to this server
map $http_x_forwarded_proto $proxy_x_forwarded_proto {
  default $http_x_forwarded_proto;
  ''      $scheme;
}
# If we receive X-Forwarded-Port, pass it through; otherwise, pass along the
# server port the client connected to
map $http_x_forwarded_port $proxy_x_forwarded_port {
  default $http_x_forwarded_port;
  ''      $server_port;
}
# If we receive Upgrade, set Connection to "upgrade"; otherwise, delete any
# Connection header that may have been passed to this server
map $http_upgrade $proxy_connection {
  default upgrade;
  '' close;
}
# Apply fix for very long server names
server_names_hash_bucket_size 128;
# Default dhparam
ssl_dhparam /etc/nginx/dhparam/dhparam.pem;
# Set appropriate X-Forwarded-Ssl header
map $scheme $proxy_x_forwarded_ssl {
  default off;
  https on;
}
gzip_types text/plain text/css application/javascript application/json application/x-javascript text/xml application/xml application/xml+rss text/javascript;
log_format vhost '$host $remote_addr - $remote_user [$time_local] '
                 '"$request" $status $body_bytes_sent '
                 '"$http_referer" "$http_user_agent"';
access_log off;
resolver 172.31.0.2;
# HTTP 1.1 support
proxy_http_version 1.1;
proxy_buffering off;
proxy_set_header Host $http_host;
proxy_set_header Upgrade $http_upgrade;
proxy_set_header Connection $proxy_connection;
proxy_set_header X-Real-IP $remote_addr;
proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
proxy_set_header X-Forwarded-Proto $proxy_x_forwarded_proto;
proxy_set_header X-Forwarded-Ssl $proxy_x_forwarded_ssl;
proxy_set_header X-Forwarded-Port $proxy_x_forwarded_port;
# Mitigate httpoxy attack (see README for details)
proxy_set_header Proxy "";
server {
	server_name _; # This is just an invalid value which will never trigger on a real hostname.
	listen 80;
	access_log /var/log/nginx/access.log vhost;
	return 503;
}
# whoami.local
upstream whoami.local {
}
server {
	server_name whoami.local;
	listen 80 ;
	access_log /var/log/nginx/access.log vhost;
	location / {
		proxy_pass http://whoami.local;
	}
}
```

Indeed, the `upstream whoami.local` block is empty.

I have reproduced this issue with the following minimal task definition. This is the information required by AWS ECS, and is based on the docker compose example here in the nginx-proxy readme.

```
{
    "executionRoleArn": null,
    "containerDefinitions": [
        {
            "dnsSearchDomains": null,
            "logConfiguration": null,
            "entryPoint": null,
            "portMappings": [
                {
                    "hostPort": 80,
                    "protocol": "tcp",
                    "containerPort": 80
                }
            ],
            "command": null,
            "linuxParameters": null,
            "cpu": 0,
            "environment": [],
            "ulimits": null,
            "dnsServers": null,
            "mountPoints": [
                {
                    "readOnly": null,
                    "containerPath": "/tmp/docker.sock",
                    "sourceVolume": "docker-sock"
                }
            ],
            "workingDirectory": null,
            "dockerSecurityOptions": null,
            "memory": 128,
            "memoryReservation": null,
            "volumesFrom": [
                {
                    "sourceContainer": "whoami",
                    "readOnly": null
                }
            ],
            "image": "jwilder/nginx-proxy",
            "disableNetworking": null,
            "essential": true,
            "links": null,
            "hostname": null,
            "extraHosts": null,
            "user": null,
            "readonlyRootFilesystem": null,
            "dockerLabels": null,
            "privileged": null,
            "name": "nginx-proxy"
        },
        {
            "dnsSearchDomains": null,
            "logConfiguration": null,
            "entryPoint": null,
            "portMappings": [],
            "command": null,
            "linuxParameters": null,
            "cpu": 0,
            "environment": [
                {
                    "name": "VIRTUAL_HOST",
                    "value": "whoami.local"
                }
            ],
            "ulimits": null,
            "dnsServers": null,
            "mountPoints": [],
            "workingDirectory": null,
            "dockerSecurityOptions": null,
            "memory": 128,
            "memoryReservation": null,
            "volumesFrom": [],
            "image": "jwilder/whoami",
            "disableNetworking": null,
            "essential": true,
            "links": null,
            "hostname": null,
            "extraHosts": null,
            "user": null,
            "readonlyRootFilesystem": null,
            "dockerLabels": null,
            "privileged": null,
            "name": "whoami"
        }
    ],
    "memory": null,
    "taskRoleArn": null,
    "family": "landon-test-1",
    "requiresCompatibilities": null,
    "networkMode": "bridge",
    "cpu": null,
    "volumes": [
        {
            "name": "docker-sock",
            "host": {
                "sourcePath": "/var/run/docker.sock"
            }
        }
    ],
    "placementConstraints": []
}
```
As I said, not all nodes failed. The difference found was that the failed nodes were running the latest version of ecs-agent (1.16.0). The node that was still working were running 1.14.3. We have temporarily solved this problem by deploying EC2 instances running amzn-ami-2017.03.c-amazon-ecs-optimized (ami-11120768). Via experiment we found that this AMI includes a sufficiently old version of ecs-agent. This is a short term solution and the underlying issue needs to be addressed.

My hope is that somebody in this forum can suggest troubleshooting steps that will help us discover the root cause of this problem. I'd be happy to perform additional data collection steps. The configuration included in this post should be sufficient for anybody wanting to reproduce this themselves.

**Comment by jwilder** [comment]
Sounds like this might be related to: https://github.com/jwilder/docker-gen/issues/263

**Comment by kelvin-acosta** [comment]
Is there any current work around to this issue or a fix soon?

**Comment by z00m1n** [comment]
@kelvin-acosta : A quick fix is to add `ECS_ENABLE_TASK_CPU_MEM_LIMIT=false` to `/etc/ecs/ecs.config` on the EC2 instance that is part of the ECS cluster and runs the AWS ECS agent container, see my comment at <https://github.com/aws/amazon-ecs-agent/issues/1119#issuecomment-357529259>. Confirmed with `EC2 AMI amzn-ami-2017.09.f-amazon-ecs-optimized [ami-eacf5d85]` with ecs-agent 1.16.1.

**Comment by jniedrauer** [comment]
Still an issue. Kind of disconcerting to me that amazon would make a change that breaks the docker spec.

**Comment by jrbenito** [comment]
This is still an issue on Amazon ECS.

@z00m1n fix might work but break one desirable feature of ECS. It might be usable or not depending on if one needs to user LIMITS on ECS.

**Comment by levic92** [comment]
In my instance on aws ecs the problem is at this line:

https://github.com/jwilder/nginx-proxy/blob/master/nginx.tmpl#L130

If I replace that block with this code then it all works fine:

```
upstream {{ $upstream_name }} {
{{ range $container := $containers }}
	{{ $addrLen := len $container.Addresses }}
	{{ $network := index $container.Networks 0 }}
	
	{{/* If only 1 port exposed, use that */}}
	{{ if eq $addrLen 1 }}
		{{ with $address := index $container.Addresses 0 }}
			# {{$container.Name}}
			server {{ $network.IP }}:{{ $address.Port }};
		{{ end }}

	{{/* If more than one port exposed, use the one matching VIRTUAL_PORT env var */}}
	{{ else if $container.Env.VIRTUAL_PORT }}
		{{ range $i, $address := $container.Addresses }}
			{{ if eq $address.Port $container.Env.VIRTUAL_PORT }}
			# {{$container.Name}}
			server {{ $network.IP }}:{{ $address.Port }};
			{{ end }}
		{{ end }}

	{{/* Else default to standard web port 80 */}}
	{{ else }}
		{{ range $i, $address := $container.Addresses }}
			{{ if eq $address.Port "80" }}
			# {{$container.Name}}
			server {{ $network.IP }}:{{ $address.Port }};
			{{ end }}
		{{ end }}
	{{ end }}
{{ end }}
}
```

**Comment by ceefour** [comment]
Thanks to @levic92's tip: https://github.com/jwilder/nginx-proxy/issues/1004#issuecomment-448019795

I'm able to run it successfully in ECS with this configuration:
my app image is bitnami/wordpress
no need to configure ports
just VIRTUAL_HOST
no need to configure DOCKER_PROVIDER

If anyone else wants it, it's available in dockerhub so you can use it: https://hub.docker.com/r/hendy/nginx-proxy

But this will get outdated really soon.

I hope @jwilder will fix this to work well with ECS without hacks.

Thank you!

**Comment by ceefour** [comment]
Bump.

Please update this, it's Aug 2019 and the problem still persists.

**Comment by nicaurybenitez** [comment]
Please edit the nginx.tmpl like  @levic92  and replace the block

upstream {{ $upstream_name }} {
{{ range $container := $containers }}
	{{ $addrLen := len $container.Addresses }}
	{{ $network := index $container.Networks 0 }}
	
	{{/* If only 1 port exposed, use that */}}
	{{ if eq $addrLen 1 }}
		{{ with $address := index $container.Addresses 0 }}
			# {{$container.Name}}
			server {{ $network.IP }}:{{ $address.Port }};
		{{ end }}

	{{/* If more than one port exposed, use the one matching VIRTUAL_PORT env var */}}
	{{ else if $container.Env.VIRTUAL_PORT }}
		{{ range $i, $address := $container.Addresses }}
			{{ if eq $address.Port $container.Env.VIRTUAL_PORT }}
			# {{$container.Name}}
			server {{ $network.IP }}:{{ $address.Port }};
			{{ end }}
		{{ end }}

	{{/* Else default to standard web port 80 */}}
	{{ else }}
		{{ range $i, $address := $container.Addresses }}
			{{ if eq $address.Port "80" }}
			# {{$container.Name}}
			server {{ $network.IP }}:{{ $address.Port }};
			{{ end }}
		{{ end }}
	{{ end }}
{{ end }}
}

**Comment by RyanGannon** [comment]
> In my instance on aws ecs the problem is at this line:
> 
> https://github.com/jwilder/nginx-proxy/blob/master/nginx.tmpl#L130
> 
> If I replace that block with this code then it all works fine:
> 
> ```
> upstream {{ $upstream_name }} {
> {{ range $container := $containers }}
> 	{{ $addrLen := len $container.Addresses }}
> 	{{ $network := index $container.Networks 0 }}
> 	
> 	{{/* If only 1 port exposed, use that */}}
> 	{{ if eq $addrLen 1 }}
> 		{{ with $address := index $container.Addresses 0 }}
> 			# {{$container.Name}}
> 			server {{ $network.IP }}:{{ $address.Port }};
> 		{{ end }}
> 
> 	{{/* If more than one port exposed, use the one matching VIRTUAL_PORT env var */}}
> 	{{ else if $container.Env.VIRTUAL_PORT }}
> 		{{ range $i, $address := $container.Addresses }}
> 			{{ if eq $address.Port $container.Env.VIRTUAL_PORT }}
> 			# {{$container.Name}}
> 			server {{ $network.IP }}:{{ $address.Port }};
> 			{{ end }}
> 		{{ end }}
> 
> 	{{/* Else default to standard web port 80 */}}
> 	{{ else }}
> 		{{ range $i, $address := $container.Addresses }}
> 			{{ if eq $address.Port "80" }}
> 			# {{$container.Name}}
> 			server {{ $network.IP }}:{{ $address.Port }};
> 			{{ end }}
> 		{{ end }}
> 	{{ end }}
> {{ end }}
> }
> ```

Was up all night chasing my tail on this one, thank you.

**Comment by buchdag** [comment]
This might has been fixed by #1601, if anyone is willing to test.

**Comment by PaulVanSchayck** [comment]
I was affected by this issue before, but I'm pretty sure this was indeed fixed. This issue could perhaps be closed.
